### PR TITLE
make cdc runnable

### PR DESCRIPTION
**Pull request by fewdan**
### What problem does this PR solve? 

run cdc on k8s

### Check List 

Tests 

 - No code

Code changes

 - Has Go code change

Side effects

 - NO

Related changes

 - NO

### Does this PR introduce a user-facing change?:

```release-note
NONE
```

### CODE REVIEW

File: cmd/cdc-pocket/main.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+	flag.Parse()
+	cfg := control.Config{
+		Mode:        control.ModeSelfScheduled,
+		ClientCount: 1,
+		RunTime:     fixture.Context.RunTime,
+		RunRound:    1,
+	}
+	pocketConfig := config.Init()
+	pocketConfig.Options.Serialize = false
+	pocketConfig.Options.Path = fixture.Context.ABTestConfig.LogPath
```
Comment (mahjonp) [inline cmd/cdc-pocket/main.go]: IMO `ABTestConfig.LogPath` seems not suitable here, could you change another flag?

File: pkg/test-infra/tests/client.go [hunk 2]
```diff
@@ -45,7 +45,7 @@ var TestClient *TestCli
 type TestCli struct {
 	Config *rest.Config
 	Cli    client.Client
-	CDC    *cdc.CdcOps
+	CDC    *cdc.Ops
```
Comment (mahjonp) [inline pkg/test-infra/tests/client.go]: Use unified name style?

File: pkg/test-infra/cdc/operation.go [hunk 3]
```diff
@@ -16,193 +16,227 @@ package cdc
...
+		timeout, st.Name)
+	return wait.PollImmediate(5*time.Second, timeout, func() (bool, error) {
+		key, err := client.ObjectKeyFromObject(local)
+		if err != nil {
+			return false, err
+		}
+
+		err = c.cli.Get(context.TODO(), key, local)
+		if err != nil && errors.IsNotFound(err) {
+			return false, err
```
Comment (mahjonp) [inline pkg/test-infra/cdc/operation.go]:
```suggestion
			return false, nil
```

File: pkg/test-infra/cdc/operation.go [hunk 4]
```diff
@@ -16,193 +16,227 @@ package cdc
...
+		}
+		if err != nil {
+			log.Errorf("error getting cdc statefulset: %v", err)
+			return false, err
+		}
+		if local.Status.ReadyReplicas != *local.Spec.Replicas {
+			log.Infof("StatefulSet %s found but there are %d ready replicas and %d total replicas.",
```
Comment (mahjonp) [inline pkg/test-infra/cdc/operation.go]:
```suggestion
			log.Infof("CDC %s do not have enough ready replicas, ready: %d, desired: %d",
```

File: pkg/test-infra/cdc/operation.go [hunk 5]
```diff
@@ -16,193 +16,227 @@ package cdc
...
+				local.Name, local.Status.ReadyReplicas, *local.Spec.Replicas)
+			return false, nil
+		}
+		log.Infof("All %d replicas of StatefulSet %s are ready.", local.Status.ReadyReplicas, local.Name)
```
Comment (mahjonp) [inline pkg/test-infra/cdc/operation.go]:
```suggestion
		log.Infof("All %d replicas of CDC %s are ready.", *local.Status. Replicas, local.Name)
```

File: pkg/test-infra/cdc/operation.go [hunk 6]
```diff
@@ -16,193 +16,227 @@ package cdc
...
+
+	return wait.PollImmediate(5*time.Second, timeout, func() (bool, error) {
+		key, err := client.ObjectKeyFromObject(local)
+		if err != nil {
+			return false, err
+		}
+
+		err = c.cli.Get(context.TODO(), key, local)
+		if err != nil && errors.IsNotFound(err) {
+			return false, err
```
Comment (mahjonp) [inline pkg/test-infra/cdc/operation.go]: ditto

File: pkg/test-infra/cdc/operation.go [hunk 7]
```diff
@@ -16,193 +16,227 @@ package cdc
...
+	}
+
+	return nil
 }
 
-func (c *CdcOps) ApplyCDC(spec *CDCSpec) (*CDC, error) {
-	if spec.Image == "" {
-		spec.Image = fixture.Context.CDCImage
+func (c *Ops) ApplyCDC(cc *CDC) error {
+	if err := c.ApplyObject(cc.StatefulSet); err != nil {
```
Comment (mahjonp) [inline pkg/test-infra/cdc/operation.go]: let it be a private method?

File: pkg/test-infra/cdc/operation.go [hunk 8]
```diff
@@ -16,193 +16,227 @@ package cdc
...
+
+	if err := c.ApplyTiDBCluster(tc.Downstream); err != nil {
+		return err
+	}
+
+	if err := c.ApplyCDC(tc.CDC); err != nil {
+		return err
+	}
+
+	if err := c.ApplyJob(tc.CDC.Job); err != nil {
```
Comment (mahjonp) [inline pkg/test-infra/cdc/operation.go]: make `ApplyJob` private?

File: pkg/test-infra/fixture/fixture.go [hunk 9]
```diff
@@ -180,6 +181,10 @@ func init() {
 	flag.StringVar(&Context.ABTestConfig.Cluster2Version, "abtest.image-version", "", "specify version for cluster B")
 	flag.StringVar(&Context.ABTestConfig.LogPath, "abtest.log", "", "log path for abtest, default to stdout")
 
+	flag.StringVar(&Context.CDCConfig.CDCVersion, "cdc-version", "", `overwrite "-image-version" flag for CDC`)
+	flag.StringVar(&Context.CDCConfig.DockerRepository, "cdc-repository", "", `specify docker registry for CDC`)
+	flag.StringVar(&Context.CDCConfig.HubAddress, "cdc-hub", "", `overwrite "-hub" flag for CDC`)
```
Comment (mahjonp) [inline pkg/test-infra/fixture/fixture.go]: group cdc flag into `cdc.xxx`?

File: pkg/test-infra/fixture/fixture.go [hunk 10]
```diff
@@ -157,6 +157,7 @@ func init() {
 	flag.StringVar(&Context.HistoryFile, "history", "./history.log", "history file record client operation")
 
 	flag.StringVar(&Context.Namespace, "namespace", "", "test namespace")
+	flag.StringVar(&Context.MySQLVersion, "mysql-version", "5.6", "Default mysql version")
```
Comment (mahjonp) [inline pkg/test-infra/fixture/fixture.go]: change mysql-version into `cdc.mysql-vesion`?

File: pkg/test-infra/cdc/operation.go [hunk 11]
```diff
@@ -16,193 +16,227 @@ package cdc
...
 
-	if err := c.cli.Create(context.TODO(), kjob); err != nil {
+//  ApplyJob applies cdc job
+func (c *Ops) ApplyJob(job *batchv1.Job) error {
+	if err := c.ApplyObject(job); err != nil {
 		return err
 	}
-
-	return nil
+	return c.waitJobCompleted(job, 5*time.Second)
```
Comment (mahjonp) [inline pkg/test-infra/cdc/operation.go]: 5s maybe too short here?

Comment (mahjonp) [reply pkg/test-infra/cdc/operation.go]: And I think it' unnecessary to parameterize it, maybe you can use `fixture.Context.WaitClusterReadyDuration` in waitJobCompleted directly.

File: pkg/test-infra/mysql/operation.go [hunk 12]
```diff
@@ -103,11 +103,9 @@ func (m *MySQLOps) renderMySQL(spec *MySQLSpec) (*MySQL, error) {
...
 	// var err error
 	if spec.Resource.Requests != nil {
-		// size := spec.Resource.Requests[fixture.Storage]
-		// q, err = resource.ParseQuantity(size)
-		// if err != nil {
-		// 	return nil, fmt.Errorf("cant' get storage size for mysql: %v", err)
-		// }
+		size := spec.Resource.Requests[fixture.Storage]
+		q = size
+		fmt.Println(q)
```
Comment (mahjonp) [inline pkg/test-infra/mysql/operation.go]: remove debug stub?

File: pkg/test-infra/fixture/fixture.go (hunk 10 above)
Comment (cwen0) [reply pkg/test-infra/fixture/fixture.go]: At present, this flag is not used by cdc.

File: pkg/cluster/k8s.go [hunk 13]
```diff
@@ -36,8 +37,8 @@ func (k *K8sProvisioner) SetUp(ctx context.Context, spec interface{}) ([]cluster
 		return k.setUpTiDBCluster(s)
 	case *binlog.Recommendation:
 		return k.setUpBinlogCluster(s)
-	case *abtest.Recommendation:
-		return k.setUpABTestCluster(s)
```
Comment (yeya24) [inline pkg/cluster/k8s.go]: Why delete abtest?

File: pkg/test-infra/cdc/operation.go [hunk 14]
```diff
@@ -16,193 +16,225 @@ package cdc
...
+	}
+
+	if err := c.applyJob(tc.CDC.Job); err != nil {
+		return err
+	}
 
-type CDC struct {
-	Sts    *appsv1.StatefulSet
-	Source *v1alpha1.TidbCluster
+	time.Sleep(2 * time.Second)
```
Comment (yeya24) [inline pkg/test-infra/cdc/operation.go]: Why sleep here? Could you please add some comments?

File: pkg/test-infra/cdc/recommander.go [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+	"github.com/pingcap/tipocket/pkg/test-infra/tidb"
+
+	appsv1 "k8s.io/api/apps/v1"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+)
+
+// CDC defines the configuration for running on Kubernetes
```
Comment (yeya24) [inline pkg/test-infra/cdc/recommander.go]: Maybe better to rename this file to `recommend`

File: pkg/test-infra/tidb/recommend.go [hunk 16]
```diff
@@ -80,6 +80,7 @@ func buildImage(name string) string {
 // RecommendedTiDBCluster does a recommendation, tidb-operator do not have same defaults yet
 func RecommendedTiDBCluster(ns, name string) *Recommendation {
 	enablePVReclaim, exposeStatus := true, true
+	always := corev1.PullAlways
```
Comment (yeya24) [inline pkg/test-infra/tidb/recommend.go]: I am not sure about this one. @fewdan Is this for testing? IMO if not present is better here.

File: pkg/test-infra/cdc/operation.go (hunk 14 above)
Comment (cwen0) [reply pkg/test-infra/cdc/operation.go]: this is unnecessary, I will remove it.

**Pull request merged**